File: src/plugin/database/BanListener.java
package plugin.database;

import arc.util.Log;
import arc.util.Strings;
import arc.util.Time;
import org.postgresql.PGConnection;
import org.postgresql.PGNotification;
import plugin.database.models.Ban;
import plugin.database.models.PlayerData;
import plugin.discord.Bot;

import java.sql.Connection;
import java.sql.Statement;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import static plugin.Bundle.sendMessage;
import static plugin.utils.Utils.formatTime;

public class BanListener {
    private static int failedTimes = 0;
    private static final ExecutorService executor = Executors.newSingleThreadExecutor();

    public static void load() {
        executor.submit(()->{
            try(Connection con = Database.dataSource.getConnection()) {
                PGConnection pgCon = con.unwrap(PGConnection.class);
                try(Statement st = con.createStatement()) {
                    st.execute("LISTEN new_ban");
                }

                while(true) {
                    PGNotification[] notifications = pgCon.getNotifications();

                    if(notifications != null)
                        for(PGNotification notify : notifications) {
                            String payload = notify.getParameter();
                            if(!Strings.canParseInt(payload)) continue;
                            int i = Strings.parseInt(payload);
                            Ban.getBan(i).ifPresent(ban->
                                    PlayerData.getPlayerById(ban.playerId).ifPresent(p-> {
                                        sendMessage("advertise.banned", ban.playerId, p.coloredName(), ban.unbanTime == null ? "never" :formatTime((ban.unbanTime.toEpochMilli() - Time.millis()) / 1000), ban.reason);
                                        ban.kickPlayer(p);
                                    })
                            );
                        }

                    Thread.sleep(1500);
                }
            } catch (Exception e) {
                Log.err(e);
                Bot.sendLog(e.getMessage());
                if(failedTimes < 5) {
                    load();
                    failedTimes += 1;
                }
            }
        });
    }
}

File: src/plugin/database/BanListener.kt
package plugin.database

import arc.util.Log
import arc.util.Strings
import arc.util.Time
import mindustry.gen.Player
import org.postgresql.PGConnection
import plugin.Bundle
import plugin.database.Database.dataSource
import plugin.database.models.getBan
import plugin.database.models.getPlayerById
import plugin.discord.Bot.sendLog
import plugin.utils.formatTime
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.function.Consumer

object BanListener {
    private
    var failedTimes: Int = 0
    private
    val executor: ExecutorService = Executors.newSingleThreadExecutor()

    fun load() {
        executor.submit {
            try {
                dataSource!!.getConnection().use { con ->
                    val pgCon = con.unwrap(PGConnection::class.java)
                    con.createStatement().use { st ->
                        st.execute("LISTEN new_ban")
                    }
                    while (true) {
                        val notifications = pgCon.notifications

                        if (notifications != null) for (notify in notifications) {
                            val payload = notify.parameter
                            if (!Strings.canParseInt(payload)) continue
                            val i = Strings.parseInt(payload)
                            getBan(i)?.let { ban ->
                                getPlayerById(ban.playerId).ifPresent(Consumer { p: Player? ->
                                    Bundle.sendMessage(
                                        "advertise.banned",
                                        ban.playerId,
                                        p!!.coloredName(),
                                        if (ban.unbanTime == null) "never" else formatTime((ban.unbanTime.toEpochMilli() - Time.millis()) / 1000),
                                        ban.reason
                                    )
                                    ban.kickPlayer(p)
                                })
                            }
                        }

                        Thread.sleep(1500)
                    }
                }
            } catch (e: Exception) {
                Log.err(e)
                sendLog(e.message)
                //if (failedTimes < 5) {
                load()
                failedTimes += 1
                //}
            }
        }
    }
}